core/Archive.php

```diff
@@ -547,8 +547,11 @@ protected function get($archiveNames, $archiveDataType, $idSubtable = null)
             $dataNames, $archiveDataType, $this->params->getIdSites(), $this->params->getPeriods(), $defaultRow = null);
 
         $archiveIds = $this->getArchiveIds($archiveNames);
-
         if (empty($archiveIds)) {
+            /**
+             * Triggered when no archive data is found in an API request.
+             */
+            Piwik::postEvent('Archive.noArchivedData');
```

Comment:
I feel like this might not be the best place for some reason. Eg it would stop archiving everything else afterwards? Eg I just recently implemented some logic to trigger an archiving of a segment within another archiving process and this segment might not be defined etc. Also there might be no ArchiveIds when there is simply no data etc? 

I feel like it might be more of an event that should be handled eg in the visualization if there is no data for a report, then check if it a segment was used and check if it will be archived or not.

The challenge be how to do this for API requests and that's why you likely added this event here. The report rendered doesn't seem like the best place either. The post processor may work for this event but not really sure either. It seems like there is no good place. 

I might have to think about this a bit longer but feeling like throwing exceptions here can cause lots of random side effects that we maybe aren't prepared for. I would love if it could be just handled in the visualization itself, and then somehow also for API requests (when the root request is an API request maybe only, not when executing an API request within a regular UI request or so)

Reply:
Maybe it would already help to do it only if eg `Root::isRootApiRequest() && !$ArchiveTrigger` or so.

Reply:
Yes I originally had it in the Visualization, and it worked for one ticket, but not for the other one. I thought about doing it on Request.dispatch for API methods but that has odd edge cases as well (eg, we'd have to check if what we got is an empty datatable, but what about datatable\maps? and what about plugins that return something else? or plugins that just do something that doesn't trigger archiving?).

Checking for root API request that isn't in cron archiving sounds better to me. This would be in the event handler, correct? Or did you mean something else?

Reply:
Yeah I totally see what you mean. I'll have another think and get back to you later (in about 1.5 hours) or at the latest tmrw morning.

Reply:
I can't think of any other place. Maybe we could tag the events with `@ignore` so they won't be public API allowing us to change things later? Then in the event we would only throw the exception if possible when we are in an actual API request (`Request::isRootApiRequest()` or similar). 

Also we need to check maybe if there it really hasn't been archived or whether there was just no data to be archived? Not sure but something to think about.

I'm thinking we actually need another method `Request::isRootApiRequestHandlingMethod($method . '.' . $action)` as we maybe don't want to throw an exception if an API request eg without a segment executes another API request with a segment (using `Request::processRequest`)? It gets quite complicated though if you know what I mean. I basically just want to avoid throwing an exception maybe when it is not "expected" to not really break any possible logic etc. But we could maybe try without this at first (but then we never add the check unless we do run into a problem).

Reply:
> Haven't looked at isRootApiRequest(), but if another request is executed inside one request, then the "sub-request" isn't the root one right?

Right, but so far we are not handling this in that method.

Reply:
Or lets say that method only checks if originally an API request is executed but not whether we are currently in that request or not. This would likely me a different method.